org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java

```diff
@@ -0,0 +1,606 @@
+
+    private Messages.Response setBody(Messages.Response response, Responses.ResponseBody body) {
+        response.body = body;
+        if (body instanceof Responses.ErrorResponseBody) {
+            response.success = false;
+            Types.Message error = ((Responses.ErrorResponseBody) body).error;
+            if (error.format != null) {
+                response.message = error.format;
+            } else {
+                response.message = "Error response body";              
+            }
+        } else {
+            response.success = true;
+            if (body instanceof Responses.InitializeResponseBody) {
+                response.body = ((Responses.InitializeResponseBody) body).body;
+            }
+        }
+        return response;
+      }
+
+    private void sendEvent(Events.DebugEvent event) {
+        this.eventConsumer.accept(event);
+    }
+
+    private Types.Source getAdapterSource(Location location) throws URISyntaxException, AbsentInformationException {
+        Types.Source source = null;
+        String uri = null;
+        String name = location.sourceName();
+        try {
+            uri = AdapterUtils.getURI(location);
+        } catch (JavaModelException e) {
+            // do nothing.
+        }
+
+        if (uri != null && uri.startsWith("jdt://")) {
+            int sourceReference = sourceCollection.create(uri);
+            source = new Types.Source(name, uri, sourceReference);
+        } else if (uri != null) {
+            source = new Types.Source(name, Paths.get(new URI(uri)).toString(), 0);
+        } else {
+            String originalSourcePath = location.sourcePath();
+            String sourcepath = AdapterUtils.sourceLookup(this.sourcePath, originalSourcePath);
+            if (sourcepath == null) {
+                sourcepath = Paths.get(this.cwd, originalSourcePath).toString();
+            }
+            source = new Types.Source(name, sourcepath, 0);
+        }
+        return source;
+    }
+
+    private ThreadReference getThread(int threadId) {
+        for (ThreadReference thread : this.safeGetAllThreads()) {
+            if (thread.uniqueID() == threadId) {
+                return thread;
+            }
+        }
+        return null;
+    }
+
+    private List<ThreadReference> safeGetAllThreads() {
+        try {
+            return this.debugSession.allThreads();
+        } catch (VMDisconnectedException ex) {
+            return new ArrayList<>();
+        }
+    }
+
+    private int convertDebuggerLineToClient(int line) {
+        if (this.debuggerLinesStartAt1) {
+            return this.clientLinesStartAt1 ? line : line - 1;
+        } else {
+            return this.clientLinesStartAt1 ? line + 1 : line;
+        }
+    }
+
+    private int convertClientLineToDebugger(int line) {
+        if (this.debuggerLinesStartAt1) {
+            return this.clientLinesStartAt1 ? line : line + 1;
+        } else {
+            return this.clientLinesStartAt1 ? line - 1 : line;
```

Comment:
Here client represents vscode, debugger represents jvm. If line number sent by vscode starts with 0, but jvm parses java file starting with 1, i think in such case it's necessary to translate vscode line number to jvm line number style.